exchange/websocket/manager.go

```diff
@@ -472,102 +479,90 @@ func (m *Manager) connect() error {
-			if err != nil {
-				multiConnectFatalError = fmt.Errorf("%s websocket: %w", m.exchangeName, common.AppendError(ErrSubscriptionFailure, err))
+		if !cW.setup.SubscriptionsNotRequired {
+			if cW.setup.Subscriber == nil {
+				fatalErr = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, cW.setup.URL, errWebsocketSubscriberUnset)
 				break
 			}
-
-			if len(subs) == 0 {
+			if len(connSubs) == 0 {
 				// If no subscriptions are generated, we skip the connection
 				if m.verbose {
-					log.Warnf(log.WebsocketMgr, "%s websocket: no subscriptions generated", m.exchangeName)
+					log.Warnf(log.WebsocketMgr, "%s websocket: no subscriptions generated [conn:%d] [URL:%s]", m.exchangeName, i+1, cW.setup.URL)
 				}
 				continue
 			}
 		}
 
-		if m.connectionManager[i].setup.Connector == nil {
-			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, m.connectionManager[i].setup.URL, errNoConnectFunc)
+		if cW.setup.Connector == nil {
+			fatalErr = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, cW.setup.URL, errNoConnectFunc)
 			break
 		}
-		if m.connectionManager[i].setup.Handler == nil {
-			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, m.connectionManager[i].setup.URL, errWebsocketDataHandlerUnset)
-			break
-		}
-		if m.connectionManager[i].setup.Subscriber == nil && !m.connectionManager[i].setup.SubscriptionsNotRequired {
-			multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, m.connectionManager[i].setup.URL, errWebsocketSubscriberUnset)
+		if cW.setup.Handler == nil {
+			fatalErr = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, cW.setup.URL, errWebsocketDataHandlerUnset)
 			break
 		}
 
 		// TODO: Add window for max subscriptions per connection, to spawn new connections if needed.
 
-		conn := m.getConnectionFromSetup(m.connectionManager[i].setup)
+		conn := m.getConnectionFromSetup(cW.setup)
 
-		if err := m.connectionManager[i].setup.Connector(context.TODO(), conn); err != nil {
-			multiConnectFatalError = fmt.Errorf("%v Error connecting %w", m.exchangeName, err)
+		if err = cW.setup.Connector(context.TODO(), conn); err != nil {
+			fatalErr = fmt.Errorf("%v Error connecting %w", m.exchangeName, err)
 			break
 		}
 
 		if !conn.IsConnected() {
-			multiConnectFatalError = fmt.Errorf("%s websocket: [conn:%d] [URL:%s] failed to connect", m.exchangeName, i+1, conn.URL)
+			fatalErr = fmt.Errorf("%s websocket: [conn:%d] [URL:%s] failed to connect", m.exchangeName, i+1, conn.URL)
 			break
 		}
 
-		m.connections[conn] = m.connectionManager[i]
-		m.connectionManager[i].connection = conn
+		m.connections[conn] = cW
+		cW.connection = conn
 
-		m.Wg.Add(1)
-		go m.Reader(context.TODO(), conn, m.connectionManager[i].setup.Handler)
+		m.Wg.Go(func() { m.Reader(context.TODO(), conn, cW.setup.Handler) })
```

Comment:
Need to rm the `defer m.Wg.Done()` in m.Reader or it will panic at the disco.

Reply:
I've fixed up the tests for it and I'm pretty happy, so I've pushed it to the branch too.

Fixed https://github.com/gbjk/gocryptotrader/commit/be82ea61c